Team Billing

jamiepine opened the pull request:
## What does this PR do?
Charge team owners on a per seat basis for all non-pro members of a team.

Issue: https://github.com/calendso/calendso/issues/1472

### How to review
- Set `HOSTED_CAL_FEATURES=true` in your env
- Ensure your env is authed with Cal's official Stripe testing account
- Sign into pro@example.com and create a team
- Add members, remove members, leave teams & witness billing update accordingly

## Test Cases
- Team with existing members without seats shows alert notice for team owner
- Team owner can purchase missing seats
- Members are upgraded to pro when added to a team
- Seat is added in background when member is added
- Members with Pro are not counted when purchasing seats
- Owner is not counted when purchasing seat
- Members on trial status are included in seat count
- Members on free status are hidden from team
- Members are downgraded when team is deleted
- Members are downgraded when removed from/leave team
- Members with their own team can purchase seats for their members while their pro sponsored by another team
- Members with their own team are prompted to upgrade to pro when removed from team sponsoring their pro

## Deployment Strategy
- Email detailing new billing system sent to all Pro members that have a team
```sql
SELECT username, email
FROM users
LEFT JOIN "Membership" m ON m."userId" = id
WHERE role = 'OWNER'::"MembershipRole"
```
- All team members (not owner) without a stripe subscription are downgraded to Trial status
```sql
UPDATE users u	
SET plan = 'TRIAL'::"UserPlan"
FROM (
    SELECT id, metadata
    FROM users
        LEFT JOIN "Membership" m ON m."userId" = id
    WHERE plan = 'PRO'::"UserPlan"
	AND metadata IS NULL
	AND m."role" != 'OWNER'::"MembershipRole"
) as usrs
WHERE  u.id = usrs.id
```
- ~Flexible pro pricing options to be created on Stripe~ ([done](https://dashboard.stripe.com/prices/price_1KHkoeH8UDiwIftkkUbiggsM))
- Ensure multiple owners can't exist for one team (check prod for legacy teams) 

### Todo:
- [x] [Backend] Seat calculator: `getTeamSeatStats` .
        - Returns object of keys `unpaidSeats` and `freeSeats`.
        - Free seats are the owner themselves and any existing pro accounts.
- [x] [Backend] Upgrade function: `upgradeTeam`.
        - Handles switching to flexible pro pricing option with quantity derived from `unpaidSeats`.
- [x] [Backend] Downgrade function: `downgradeTeamMembers`.
        - Fired when Stripe reports failed billing for team owner's pro account
- [x] Notice + CTA to purchase missing seats for existing members.
        - Placed on team settings page ~and/or at bottom of sidebar~.
- [x] Notice when adding member that it will increase the team's seat count.
        - A modal with information about how the seat count will be increased and CTA to continue.
- [x] Increase Stripe subscription quantity when new member is added.
- [x] Decrease Stripe subscription quantity when a member is removed.
- [x] A notice that the team owner must upgrade, or that they can purchase a pro license.
        - Team owner has 14 days to upgrade while team members are switched to trial plan.
        - ~Potentially notice on event type page.~
- [x] Badge with tooltip on team list.
- [x] Hide free users from public team page
- [x] Remove "Create Event Type" button for non admins
- [x] Fix padding issue with availability button on MemberListItem
- [x] HOSTED_CAL_FEATURES to enable team billing for testing, defaults to TRUE in prod 

Billing could be discounted at scale, for example:
- $12 per month per user for first 10 users
- $10 per month per user for each additional user for first 100 users
- $8 per month per user for each additional user
This however will be configurable via Stripe depending on the business department. 

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update


## Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code and corrected any misspellings
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

jamiepine committed:
added base logic for team billing

- moved Stripe customer related logic to customer.ts
- implemented unstable logic for team owner upgrading, downgrading and adding/removing seats

PeerRich committed:
Merge branch 'main' into feature/team-billing

jamiepine committed:
logic improvements

PeerRich committed:
Merge branch 'main' into feature/team-billing

jamiepine committed:
- improved Alert style
- hide free team members on public team page
- upgraded textarea to ui component TextArea in SAML setup
- added Alert on team settings for hidden members
- hide CreateEventTypeButton if not admin
- fixed missing locale strings in team settings

jamiepine committed:
remove random import

jamiepine committed:
- show hidden status on team list
- refactor team pill

jamiepine committed:
- improved logic (mostly functional)
- added Alerts for members & owners
- added local strings
- created upgrade modal
- added info notice on invite member modal
- fixed router redirect after leaving team

jamiepine committed:
- improved logic in team-billing
- error display on upgrade modal
- added better launch.json for VSCode debugger
- fixed bug with missing inviteeUserId

jamiepine committed:
code cleanup

jamiepine committed:
nit pick fixes i should sleep now

jamiepine committed:
fixed leave team bug
- quantity would not decrease upon leave or removal

jamiepine committed:
added stripe billing callback handler

jamiepine committed:
- better launch.json
- teams empty component

jamiepine committed:
- fixed error not removing after successful pro upgrade
- fixed silent fail on team create name conflict
- fixed input border radius on member invite modal

jamiepine committed:
updated local strings

jamiepine committed:
improved logic for edge cases, such as:
- team owned by member sponsored by another team can smoothly upgrade to pro if kicked from sponsored team
- logic to calculate if owner is specifically missing pro subscription (ownerIsMissingSeat)
- corrected calculation of members missing seats, shouldn't care for proPaidForByTeamId as that only matters for removing member and preserving pro if they pay for it themselves
- added react query devtools
- added missing locale string

jamiepine committed:
Merge branch 'main' into feature/team-billing

PeerRich committed:
Merge branch 'main' into feature/team-billing

jamiepine committed:
- allow type override for LinkIconButton
- consolidate filter logic for getMembersMissingSeats

jamiepine committed:
Merge remote-tracking branch 'origin/feature/team-billing' into feature/team-billing

jamiepine committed:
- only activate team billing for hosted cal
- fix prod price keys

jamiepine committed:
Merge branch 'main' into feature/team-billing

jamiepine committed:
fix requiresUpgrade when not hosted by cal

jamiepine committed:
added HOSTED_CAL_FEATURES

PeerRich committed:
Merge branch 'main' into feature/team-billing

kodiakhq[bot] committed:
Merge branch 'main' into feature/team-billing

kodiakhq[bot] committed:
Merge branch 'main' into feature/team-billing

PeerRich commented:
@zomars can you go through this with @jamiepine once you are back? Pretty urgent and its good that you two should discuss all of the edge-cases

kodiakhq[bot] committed:
Merge branch 'main' into feature/team-billing

PeerRich commented on ee/pages/api/integrations/stripepayment/portal.ts:
```diff
@@ -15,29 +15,7 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
+    const customerId = await getStripeCustomerFromUser(session.user.id);
 
     const return_url = `${process.env.BASE_URL}/settings/billing`;
     const stripeSession = await stripe.billingPortal.sessions.create({
```
`customer: customerId,` throws a type error, which likely causes some side effects given the types come from stripe and are likely up to date, no?
<img width="999" alt="image" src="https://user-images.githubusercontent.com/8019099/152697278-6bbf477e-3be7-4ffb-af78-ed98a1c3beb1.png">

PeerRich replied:
looking at git-blame, i believe this was introduced earlier by @zomars

jamiepine replied:
fixed, was just expecting a non null value so I added a null check

kodiakhq[bot] committed:
Merge branch 'main' into feature/team-billing

kodiakhq[bot] committed:
Merge branch 'main' into feature/team-billing

kodiakhq[bot] committed:
Merge branch 'main' into feature/team-billing

kodiakhq[bot] committed:
Merge branch 'main' into feature/team-billing

pumfleet reviewed (commented):
Code looks good to me, will test further

kodiakhq[bot] committed:
Merge branch 'main' into feature/team-billing

kodiakhq[bot] committed:
Merge branch 'main' into feature/team-billing

jamiepine committed:
fixed failing build

- fixed broken import path
- added support for premium price plan. (will consider premium as a valid seat)
- remove rouge console log

jamiepine committed:
fix customer id type error

jamiepine commented on pages/team/[slug].tsx:
```diff
@@ -26,6 +27,8 @@ function TeamPage({ team }: TeamPageProps) {
   const showMembers = useToggleQuery("members");
   const { t } = useLocale();
 
+  console.log({ team });
```
console log neutralized

kodiakhq[bot] committed:
Merge branch 'main' into feature/team-billing

PeerRich commented:
awesome stuff, waiting for tomorrow to run the query with a few more eyes
